internal/registry/inmemory.go

```diff
@@ -0,0 +1,188 @@
+package registry
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/internal/signal"
+	pb "github.com/pomerium/pomerium/pkg/grpc/registry"
+
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type inMemoryServer struct {
+	ttl time.Duration
+	// onchange is used to broadcast changes to listeners
+	onchange *signal.Signal
+	// regs is {service,endpoint} -> expiration time mapping
+	regs map[inMemoryKey]*timestamppb.Timestamp
+	// mu holds lock for regs
+	mu sync.RWMutex
+}
+
+type inMemoryKey struct {
+	kind     pb.ServiceKind
+	endpoint string
+}
+
+// NewInMemoryServer constructs new registry tracking service that operates in RAM
+// as such, it is not usable for multi-node deployment where REDIS or other alternative should be used
+func NewInMemoryServer(ctx context.Context, ttl time.Duration) pb.RegistryServer {
+	srv := &inMemoryServer{
+		ttl:      ttl,
+		regs:     make(map[inMemoryKey]*timestamppb.Timestamp),
+		onchange: signal.New(),
+	}
+	go srv.periodicCheck(ctx)
+	return srv
+}
+
+func (s *inMemoryServer) periodicCheck(ctx context.Context) {
+	after := s.ttl * purgeAfterTTLFactor
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info().Msg("grpc.service_registry.PeriodicCheck/Stop")
+			return
+		case <-time.After(after):
```

Comment:
should probably use a `time.Ticker`

Reply:
as discussed, that's very minor improvement vs need for manual resource management